### ISSUE TITLE
Database migration error going from 7.8.3 to 7.9.1

### ISSUE THREAD

**Issue by bg-govau** [concourse/concourse#8692]
## Summary

After upgrading a concourse deployment from 7.8.3 to 7.9.1 we found that `concourse web` would not start due to a database migration error.

## Steps to reproduce

I have not been able to reproduce. We were able to successfully upgrade four other concourse deployments without hitting this issue.

## Expected results

`concourse web` to start successfully

## Actual results

`concourse web` did not start. We saw this in stderr every time it tried to start:

> error: failed to connect to database: migration '1653924132_int_to_bigint.up.sql' failed and was rolled back: pq: could not create unique index "jobs_unique_pipeline_id_name"

## Additional context

We deploy concourse with bosh using https://github.com/concourse/concourse-bosh-deployment

We were able to roll back the upgrade in our pipeline by re-running the previous deploy concourse build that succeeded using "re-run with the same inputs" to restore this affected concourse deployment back to 7.8.3, and `concourse web` is now running happily again.

At the same time as the upgrade to 7.9.1 on all deployments we were also:
- enabling forwarding build logs to syslog.
- redact_secrets was previously not set, now it is explicitly set to false (we assumed this would be a noop)

Here is the relevant parts of the diff shown by `bosh deploy` :
```
  instance_groups:
  - name: web
    jobs:
    - name: web
      release: concourse
      properties:
+       redact_secrets: false
+       syslog:
+         address: 127.0.0.1:514
+         hostname: concourse-atc-build-logs
+         transport: tcp
```

## Triaging info

* Concourse version: 7.9.1
* Did this used to work? Yes, and our other concourse deployments were able to be upgraded to 7.9.1 successfully.

**Comment by xtremerui** [comment]
Is there a way you can fetch the postgresql DB log where `could not create unique index "jobs_unique_pipeline_id_name"` happens? I suspect there is some corrupted data.

**Comment by bg-govau** [comment]
Good tip, I think I've found a duplicate entry in the `jobs` table which violates that constraint... Im not sure how that would be possible to happen in a live system if that constraint is there as part of the initial schema... I'll keep investigating tomorrow anyway.

**Comment by bg-govau** [comment]
The duplicate entry in the jobs table had a `name` of "foo-bar_v2". Both rows had `active = 'f'`, I think that means the job has been deleted from the current version of the pipeline.

Our theory is it this deployment has been impacted by this locale change issue in postgres when upgrading glibc - https://wiki.postgresql.org/wiki/Locale_data_changes

> What to do
> When an instance needs to be upgraded to a new glibc release, for example to upgrade the operating system, then after the upgrade
> - All indexes involving columns of type text, varchar, char, and citext should be reindexed before the instance is put into production.

We also deploy postgres using bosh, and when we went from ubuntu 18.04 bionic to 22.04 jammy stemcell we were not aware of this locale issue so we did not explicitly do a reindex after that upgrade - we upgraded to jammy around october 2022.

So it appears this meant things that relate to indexes such as checking this constraint were/maybe still are broken when dealing with strings where the locale is important - such as the hyphen in this particular job name - and that's how this bad data got in to the jobs table that violates the constraint... When the concourse upgrade to v7.9.1 modified the schema, it seems it triggered an index rebuild which was what gave us that error message due to the bad data when it realised the constraint was being violated.

We've now deleted one of the duplicate rows, and will wait a day and then try the upgrade again. I'll report back on how it goes. Happy to close this if its not a bug with concourse, but curious to know if anyone else is similarly impacted...
